content/en/blog/2023-01-10-building-a-raft-part-2.md

```diff
@@ -0,0 +1,100 @@
+draft: false
+image: img/blog/2023-01-10-building-a-raft-part-2/otterRaftTwo.png
+author: "Daniel Sollis"
+category: "Distributed Systems"
+profile: img/team/daniel-sollis.png
+description: "In the second part of our series on the Raft consensus algorithm we'll be covering log replication."
+---
+
+In our first installment on how to implement the famous Raft algorithm, we introduced the log: an ordered sequence of operations that serves as the core data structure for distributed consensus. In this second part, we'll demonstrate the beating heart of Raft, which is log replication.
+
+<!--more-->
+
+If you're following along at home, don't forget to check out this [handy Github repository containing a Raft implemention](https://github.com/rotationalio/Raft) done as an exercise for Rotational's internal Distributed Systems Bootcamp. Be sure to check out the [first post in the series](https://rotational.io/blog/building-a-raft-part-1/) too!
+
+## The Raft RPCs
+
+Raft can be implemented in three remote procedure calls (RPCs):
+
+```proto
+service Raft {
+    rpc AppendEntries(stream AppendEntriesRequest) returns (AppendEntriesReply) {};
+    rpc RequestVote(VoteRequest) returns (VoteReply) {};
+    rpc Submit(SubmitRequest) returns (SubmitReply) {};
+}
+```
+
+In the next post, we'll explore `RequestVote` and `Submit`, so in this post we'll focus on the `AppendEntries` RPC, which expects an `AppendEntriesRequest` and returns an `AppendEntriesReply`. As we'll see in the next section, these messages enable Raft's core state replication functionality.
+
+## Log Replication
+
+Log replication refers to the way we keep our peers' state synchronized in a distributed system. In Raft, we can implement log replication using an RPC called `AppendEntries`. The [Raft paper](https://web.stanford.edu/~ouster/cgi-bin/papers/raft-atc14) delineates five steps that must be executed within the `AppendEntries` RPC to ensure that log replication is safe and correct, meaning we can algorithmically prove that all peers will execute all of the same commands in the same order every time.
+
+In the next few sections, we will spend some time going over each of the 5 steps of the log replication process and their purpose. For a full implementation of the log replication function at work, check out [the AppendEntries function](https://github.com/rotationalio/Raft/blob/8a7f0417aaeafb6d301c4f5c3d57df63268c4886/raft/handlers.go#L215-L269) from the example code.
+
+## AppendEntries Steps
+
+First, let's examine the RPC definitions for the `AppendEntriesRequest` and `AppendEntriesReply` messages.
+
+When a presumptive leader (which we'll call Alpha) wants to propose a new log entry to its peers (let's call them Beta and Gamma), it will send out an `AppendEntriesRequest`. In our first post in this series, we explained that a **Term** describes a leader's continuous duration during the consensus process. Alpha will include it's perspective of the term in this `AppendEntriesRequest`:
```

Comment:
```suggestion
When a presumptive leader (which we'll call Alpha) wants to propose a new log entry to its peers (let's call them Beta and Gamma), it will send out an `AppendEntriesRequest`. In our first post in this series, we explained that a **Term** describes a leader's continuous duration during the consensus process. Alpha will include its perspective of the term in this `AppendEntriesRequest`:
```